File: modules/navigation-module/ios/NavigationModule.swift
import ExpoModulesCore

struct CoordinatesRecord: Record {
  @Field var latitude: Double = 0
  @Field var longitude: Double = 0
}

enum NavigationMode: String, Enumerable {
  case driving
  case walking
  case cycling
  case drivingTraffic = "driving-traffic"
}

public class NavigationModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NavigationModule")

    View(NavigationModuleView.self) {
      Events("onArrived", "onCancel")

      Prop("origin") { (view: NavigationModuleView, origin: CoordinatesRecord) in
        view.origin = origin
      }

      Prop("destination") { (view: NavigationModuleView, destination: CoordinatesRecord) in
        view.destination = destination
      }

      Prop("mode") { (view: NavigationModuleView, mode: NavigationMode) in
        view.mode = mode
      }

      OnViewDidUpdateProps { view in
        view.configureNavigation()
      }
    }
  }
}

File: modules/navigation-module/android/src/main/java/expo/modules/navigationmodule/NavigationModule.kt
package expo.modules.navigationmodule

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class CoordinatesRecord : Record {
  @Field
  val latitude: Double = 0.0

  @Field
  val longitude: Double = 0.0
}

class NavigationModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NavigationModule")

    View(NavigationModuleView::class) {
      Events("onArrived", "onCancel")

      Prop("origin") { view: NavigationModuleView, origin: CoordinatesRecord ->
        view.origin = origin
      }

      Prop("destination") { view: NavigationModuleView, destination: CoordinatesRecord ->
        view.destination = destination
      }

      Prop("mode") { view: NavigationModuleView, mode: NavigationMode ->
        view.mode = mode
      }

      OnViewDidUpdateProps { view ->
        view.configureNavigation()
      }
    }
  }
}
